### PR TITLE
Implement max_concurrency in SessionController

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This implements the `max_concurrency` buckets in a new `ConcurrentSessionController` class (new default). The concurrency means that multiple shards can connect simultaneously, without hitting the identify rate limit. This is very important for larger bots which can connect 16 or 64 shards at once to reduce startup time.

The buckets work by using a modulo over the `shard_id`:

```
bucket = shard_id % concurrency
```

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import com.neovisionaries.ws.client.OpeningHandshakeException;
+import net.dv8tion.jda.api.JDA;
+
+import javax.annotation.Nonnull;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+
+public class ConcurrentSessionController extends SessionControllerAdapter implements SessionController
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java:28]: Missing docs here

File: src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            try
+            {
+                node = queue.remove();
+                log.debug("Running connect node for shard {}", node.getShardInfo());
+                node.run(false); // we don't use isLast anymore because it can be a problem with many reconnecting shards
+            }
+            catch (NoSuchElementException ignored) {/* This means the node was removed before we started it */}
+            catch (InterruptedException e)
+            {
+                queue.add(node);
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java:142]: We are readding the node to the queue but still throwing the exception. This exception will kill the thread.
Are we just relying on the next call to `enqueue` to restart the thread?

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java:142]: Yes. The thread can only be interrupted by a third-party so it must be either intentional or a bug. Ideally, the third-party will also handle starting a new one in that case.

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java:142]: Okay, just wanted to make sure we were on the same page. 👍

File: src/main/java/net/dv8tion/jda/api/utils/SessionControllerAdapter.java [hunk 3]
```diff
@@ -104,8 +141,10 @@ public void handleResponse(Response response, Request<ShardedGateway> request)
 
                         String url = object.getString("url");
                         int shards = object.getInt("shards");
+                        int concurrency = object.getObject("session_start_limit")
+                                                .getInt("max_concurrency", 1);
```
Comment (GoldRenard) [inline src/main/java/net/dv8tion/jda/api/utils/SessionControllerAdapter.java]: This is missing in master after release merge. Is that ok?

Comment (GoldRenard) [reply src/main/java/net/dv8tion/jda/api/utils/SessionControllerAdapter.java]: So it means that actual `max_concurrency` is never set to SessionController now in master branch and always equals 1.

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/api/utils/SessionControllerAdapter.java]: Fixed in bdcece1

**Pull request merged**